cmd/faFindFast/efficient.go

```diff
@@ -4,133 +4,210 @@ import (
+	var gapOpenCloseFirstQuery, gapOpenedSecondQuery, gapClosedSecondQuery, numFirstQueryNs, numSecondQueryNsGap, numSecondQueryNsMatch, numSubst int // ints we will get back when moving the window one ref base.
 	var err error
 	var percentDiverged, rawPValue float64
 
 	// this caches map[k] to -log10(BinomialDist(n, k, p, true)), which is the -log10 p Value.
 	var scorePValueCache map[int]float64
 
-	// divergenceRate = -1 is a reserved value that signifies that the user has not set a divergence rate. If divergenceRate != -1,
+	// create outFile
+	file := fileio.EasyCreate(s.OutFile)
+
+	// divergenceRate = math.MaxFloat64 is a reserved value that signifies that the user has not set a divergence rate. If divergenceRate != -1,
 	// we initialize the scorePValueCache.
-	if divergenceRate != -1 {
-		scorePValueCache = binomialDistCacheLog10(windowSize, divergenceRate)
+	if s.DivergenceRate != math.MaxFloat64 {
+		scorePValueCache = binomialDistCacheLog10(s.WindowSize, s.DivergenceRate)
 	}
 
-	for lastAlnIdxOfWindow < len(reference) { // this check could also be "!done", I am not sure what is more clear
+	for lastAlnIdxOfWindow < len(firstQuery) { // this check could also be "!done", I am not sure what is more clear
 		// we always move the lastBaseOfTheWindow (right side) and add on what we find to the counters because
 		// all this stuff is now inside the current window
-		lastAlnIdxOfWindow, gapOpenCloseRef, gapOpenQuery, _, numRefNs, numQueryNsGap, numQueryNsMatch, numSubst = incrementWindowEdge(reference, query, lastAlnIdxOfWindow)
-		lastRefIdxOfWindow++
-		totalGaps += gapOpenCloseRef + gapOpenQuery
-		totalNs += numRefNs + numQueryNsGap + numQueryNsMatch
+		lastAlnIdxOfWindow, gapOpenCloseFirstQuery, gapOpenedSecondQuery, _, numFirstQueryNs, numSecondQueryNsGap, numSecondQueryNsMatch, numSubst = incrementWindowEdge(firstQuery, secondQuery, lastAlnIdxOfWindow)
+		lastFirstQueryIdxOfWindow++
+		totalGaps += gapOpenCloseFirstQuery + gapOpenedSecondQuery
+		totalNs += numFirstQueryNs + numSecondQueryNsGap + numSecondQueryNsMatch
 		totalSubst += numSubst
 
-		// usually increment the baseBeforeWindow, but not at the beginning when we have not yet incremented the end
-		// enough to have a full "windowSize" of bases in the window
-		if lastRefIdxOfWindow-refIdxBeforeWindow > windowSize {
-			alnIdxBeforeWindow, _, _, _, numRefNs, _, numQueryNsMatch, numSubst = incrementWindowEdge(reference, query, alnIdxBeforeWindow)
-			refIdxBeforeWindow++
+		// usually increment the baseBeforeWindow,
+		// but not at the beginning when we have not yet incremented the end enough to have a full "windowSize" of bases in the window
+		if lastFirstQueryIdxOfWindow-firstQueryIdxBeforeWindow > s.WindowSize {
+			alnIdxBeforeWindow, _, _, _, numFirstQueryNs, _, numSecondQueryNsMatch, numSubst = incrementWindowEdge(firstQuery, secondQuery, alnIdxBeforeWindow)
+			alnIdxBeforeWindowForRef = updateAlnIdxBeforeWindow(firstQuery, alnIdxBeforeWindow)
```

Comment:
first main new function in this PR: updateAlnIdxBeforeWindow

Reply:
I wonder if this is the major slowdown. For example, there may be big gaps in real sequences and it will take a long time to find the next nonGap base?